gateway: No hasPrefix check in ListObjectsV1 API

krisis opened the pull request:
Fixes #4411 

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
We remove the hasPrefix(marker, prefix) check from gateway's
ListObjectsV1 API delegating it to respective gateway backends (like
azure, gcs and s3).

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
See #4411 
## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Manually using sample code snippet here - https://github.com/minio/minio/issues/4411#issuecomment-303645129
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

nl5887 committed:
Implement gateway support Google Cloud Storage

nl5887 committed:
Move anonymous error to object translation from Azure specific to gateway

nl5887 committed:
Add comment, gateway should validate object name

nl5887 committed:
Added AllAccessDisabled error

nl5887 committed:
Initial implementation of Google Cloud Storage

nl5887 committed:
Add usage for GCS gateway, closes #4280

nl5887 committed:
Use default endpoint when not supplied

nl5887 committed:
Translate gcs errors to S3 compatible errors, fixes #4278

nl5887 committed:
Allow bucket creation in different regions, closes #4287 and #4241

* I needed to remove the region check from PutBucketHandler

nl5887 committed:
Return BucketAlreadyExists when bucket exists with another user

nl5887 committed:
Make every backend responsible for parsing its own arguments, fixes #4293

nl5887 committed:
Map S3 regions to Google (multi)regions

nl5887 committed:
Return correct error when PutObject fails

nl5887 committed:
Implement bucket policies

nl5887 committed:
Add access and secret key to example, needed to access Minio Gateway

nl5887 committed:
Use MINIO_REGION environment variable for region configuration, closes #4287

nl5887 committed:
Support marker, closes #4286

nl5887 committed:
Use maxKeys for iterator

nl5887 committed:
Verify md5 content hash, closes #4285

nl5887 committed:
Add region to gcs gateway example

nl5887 committed:
Support iterating through ListObjectParts using NextPartNumberMarker, closes #4284

nl5887 committed:
Verify multipart etag during complete, closes #4288

nl5887 committed:
Fix issue with UNSIGNED payloads.

Additionally also fixes escaping slashes in
temporary multipart names

nl5887 committed:
Fix issue with AbortMultipartUpload, closes #4322

nl5887 committed:
Fix uploadIDMarker handling.

nl5887 committed:
Fix ListObjectParts to list properly all parts - closes #4322

nitisht committed:
Fixes https://github.com/minio/minio/issues/4320 (#4332)

- Add description for error ErrBucketAlreadyExists

vadmeste committed:
gateway: Use default params when no args provided (#4315)

For S3 & Azure, use default parameters when no arguments (endpoint) are
provided. This also avoids a crash.

harshavardhana committed:
gateway/gcs: send proper error responses for Get/SetBucket policies. (#4338)

Fixes #4323

nl5887 committed:
Map only default region us-east-1 to gcs us region

nl5887 committed:
Cleanup and update the PR with the master branch.

Commit:
Change md5Sum to etag (#4399)

poornas committed:
Remove profile option for gcs from gateway help message (#4421)

vadmeste committed:
fix: Check project id before starting gateway (#4412)

krishnasrinivas reviewed (changes requested):
https://github.com/minio/minio/issues/4411 is updated with the fix we decided
```
for ListV1, in the response we will set marker as the continuation token returned by GCS just like how we do today, but with small change, we'll prefix the token with ##minio
This way if the client sends a marker with this prefix we know that we need to use it as token for GCS List
If the marker does not have ##minio prefix we'll List the GCS from the beginning and wait till we reach the marker so that we can send the client appropriate response

For ListV2, the implementation is more straight forward as both S3 and GCS will operate on a token. However if the client sends start-after then we will list from the server to filter out the entries till that entry.
```

poornas committed:
Set default ETag value if vendor returns empty md5 string (#4409)

The ETag is constructed from md5 atttribute of object attributes
returned by the vendor's Composer. The md5 attribute comes back
as nil for large uploads. Instead the CRC32C should be used.

Refer to https://cloud.google.com/storage/docs/hashes-etags

Fixes #4397

krisis committed:
GCS gateway allows apps to supply their own marker

Most s3 compatible apps use object keys returned in listing as
marker. This change allows this behaviour with gateway-gcs too.
